cborg/src/Codec/CBOR/ByteArray/Internal.hs

```diff
@@ -0,0 +1,83 @@
+{-# LANGUAGE BangPatterns        #-}
+{-# LANGUAGE CPP                 #-}
+{-# LANGUAGE MagicHash           #-}
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE UnboxedTuples       #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- |
+-- Module      : Codec.CBOR.ByteArray.Internal
+-- Copyright   : (c) Ben Gamari 2017-2018
+-- License     : BSD3-style (see LICENSE.txt)
+--
+-- Maintainer  : duncan@community.haskell.org
+-- Stability   : experimental
+-- Portability : non-portable (GHC extensions)
+--
+-- Various bytearray utilities
+--
+module Codec.CBOR.ByteArray.Internal
+  ( foldrByteArray
+  , copyToAddr
+  , isTrue#
+  , sameByteArray
+  , mkByteArray
+  , isByteArrayPinned
+  , touch
+  ) where
+
+import Control.Monad.ST
+import Control.Monad
+import GHC.IO (IO(..))
+import GHC.Exts
+import GHC.Word
+
+import qualified Data.Primitive.ByteArray as Prim
+
+foldrByteArray :: (Word8 -> a -> a) -> a
+               -> Int             -- ^ offset
+               -> Int             -- ^ length
+               -> Prim.ByteArray  -- ^ array
+               -> a
+foldrByteArray f z off0 len ba = go off0
+  where
+    go !off
+      | off == len = z
+      | otherwise  =
+        let x = Prim.indexByteArray ba off
+        in f x (go (off+1))
+
+copyToAddr :: Prim.ByteArray -> Int -> Ptr a -> Int -> IO ()
+copyToAddr (Prim.ByteArray ba) (I# off) (Ptr addr) (I# len) =
+    IO (\s -> case copyByteArrayToAddr# ba off addr len s of
+                s' -> (# s', () #))
+
+#if __GLASGOW_HASKELL__ < 706
+isTrue# :: Bool -> Bool
+isTrue# = id
+#endif
+
+sameByteArray :: Prim.ByteArray -> Prim.ByteArray -> Bool
+sameByteArray (Prim.ByteArray ba1#) (Prim.ByteArray ba2#) =
+    case reallyUnsafePtrEquality# (unsafeCoerce# ba1# :: ()) (unsafeCoerce# ba2# :: ()) of
+      r -> isTrue# r
+
+mkByteArray :: Int -> [Word8] -> Prim.ByteArray
+mkByteArray n xs = runST $ do
```

Comment:
There's an unstated assumption here that `n` is exactly the length of `xs`, otherwise it either leaves uninitialised memory or indexes out of bounds.

Reply:
Indeed; I suppose this does deserve a comment.